### ISSUE TITLE
Write on CAN every time a specific message is received

### ISSUE THREAD

**Issue by hayk57** [codeljo/AA_MCP2515#2]
Hello,

I have an issue when I'm trying to send a can message every time I recieve a specific message on a specific ID

for example on ID 320(DEC) 140(HEX), if data[1] == 1 I want to write a specific message on canbus, but it is blocking on the CAN.write and not moving forward

```
uint8_t dataX[8];
uint8_t toSendX[] = {0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

void onReceive(CANController&, CANFrame frame) {
  //frame.print("RX");
  id = frame.getId();
  if (id == 320)
  {
    //frame.print("RX");
    frame.getData(dataX, sizeof(dataX));
    Serial.println(dataX[1]);
    if (dataX[1] == 1){
        CANFrame frametoSendX(0X655, toSendX, sizeof(toSendX));
        CAN.write(frametoSendX); 
    }

.....
```

If someone has an advice please

Thanks

**Comment by codeljo** [comment]
Hi there, Any chance you could post a complete example that demos this issue and details on the hardware setup? Thanks

**Comment by hayk57** [comment]
I am using a Arduino Nano Every, exactly connected like this:

![image](https://github.com/codeljo/AA_MCP2515/assets/26463996/8603ed3f-3843-408e-a4af-1884cf7298fe)

I don't have any issue sending can messages, or reading with the examples you provided but when I try to send if a specific message is received, I am having issue

This is an example that is not working, I tired with filters also:

```
#include "AA_MCP2515.h"

// TODO: modify CAN_BITRATE, CAN_PIN_CS(Chip Select) pin, and CAN_PIN_INT(Interrupt) pin as required.
const CANBitrate::Config CAN_BITRATE = CANBitrate::Config_8MHz_500kbps;
const uint8_t CAN_PIN_CS = 10;
const int8_t CAN_PIN_INT = 2;

CANConfig config(CAN_BITRATE, CAN_PIN_CS, CAN_PIN_INT);
CANController CAN(config);


uint8_t dataX[8];
uint8_t toSendX[] = {0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
uint32_t id;

void onReceive(CANController&, CANFrame frame) {
  //frame.print("RX");
  id = frame.getId();
  if (id == 320)
  {
    //frame.print("RX");
    frame.getData(dataX, sizeof(dataX));
    Serial.println(dataX[1]);
    if (dataX[1] == 1){
        CANFrame frametoSendX(0X655, toSendX, sizeof(toSendX));
        CAN.write(frametoSendX); 
    }
  }
}

void onWakeup(CANController& controller) {
  controller.setMode(CANController::Mode::Normal);
}

void setup() {
  Serial.begin(115200);

  while(CAN.begin(CANController::Mode::Normal) != CANController::OK) {
    Serial.println("CAN begin FAIL - delaying for 1 second");
    delay(1000);
  }
  Serial.println("CAN begin OK");
  
  CAN.setInterruptCallbacks(&onReceive, &onWakeup);
}

void loop() {
  delay(2000);
}
```

Thanks for your support

**Comment by hayk57** [comment]
I did everything in the loop at the end

**Comment by codeljo** [comment]
> I did everything in the loop at the end

Just to confirm: You got it working by moving the write to the main loop() ?

**Comment by hayk57** [comment]
Yes, everything is in the loop, using filters to retrieve only some messages and CAN.read(frame) in the loop. I don't think this is the cleanest way of doing it.

I read the frame, compare its ID with the one I want, and if it matches, I write a specific CAN message on the bus.

**Comment by codeljo** [comment]
I just released v1.0.4 which should resolve this issue. The update should be available in the IDE within ~24 hours. Thanks for all the details as it was a big help.

**Comment by codeljo** [comment]
CANController write would block if interrupts were disabled. Version 1.0.4 resolves this issue.
